providers/implementations/ciphers/ciphercommon_gcm.c

```diff
@@ -143,20 +144,23 @@ static int setivinv(PROV_GCM_CTX *ctx, unsigned char *in, size_t inl)
 int ossl_gcm_get_ctx_params(void *vctx, OSSL_PARAM params[])
 {
     PROV_GCM_CTX *ctx = (PROV_GCM_CTX *)vctx;
+    OSSL_PARAM_IDX idx;
     OSSL_PARAM *p;
     size_t sz;
 
-    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_IVLEN);
+    ossl_scan_params_for_idx(params, &idx);
+
+    p = ossl_param_idx(params, &idx, PIDX_CIPHER_PARAM_IVLEN);
     if (p != NULL && !OSSL_PARAM_set_size_t(p, ctx->ivlen)) {
         ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
         return 0;
     }
-    p = OSSL_PARAM_locate(params, OSSL_CIPHER_PARAM_KEYLEN);
+    p = ossl_param_idx(params, &idx, PIDX_CIPHER_PARAM_KEYLEN);
```

Comment:
This...  was a surprising development.  What it does is basically to maintain the lookup idea...  but doesn't that strike you as a lot of extra churn?  See, I would have expected to see a loop here with a switch statement inside, something like this:

``` C
    for (p = params; p->key != NULL; p++) {
        switch (ossl_param_find_idx(p->key)) {
        case NIDX_CIPHER_PARAM_IVLEN:
            if (!OSSL_PARAM_set_size_t(p, ctx->ivlen)) {
                ERR_raise(ERR_LIB_PROV, PROV_R_FAILED_TO_SET_PARAMETER);
                return 0;
            }
            break;
        case NIDX_CIPHER_PARAM_KEYLEN:
            ...
```

Reply:
The PR looks reasonable to me. The use of trie matching is neat and a good design.

We could do a for-switch, and I also think that's potentially a reasonable design. If we do that, I'd expect it to iterate through the index structure, not the array of `OSSL_PARAM`. So basically a variable-length array (PIDX, OSSL_PARAM *), backed by the stack with a variable tracking the size of the array. The advantage of this is that the size of the array needed would then grow with the number of parameters supported by a function (and could either vary by function or be fixed by the upper bound of all functions we implement), rather than every parameter name we support in the entire codebase.

I'd potentially be interested in seeing an alternative PR and comparing the two approaches, though just a thought.

I think iterating through OSSL_PARAM directly would be less desirable as a solution.

The idea I had [is here](https://gist.github.com/hlandau/ec6a7438b8ab240a5bcdbb764a11db07) for completeness, but I'm just linking that for posterity.

Reply:
The disadvantage with a dynamically growing array of indexes is that you have to store it somewhere, and need to take threads and locking into account.
Your idea, @hlandau, is a neat hack ...  but it worries me that it essentially means that a provider would go and tamper directly with the application's (in this regard, libcrypto is also the application) OSSL_PARAM arrays.

Reply:
What I was suggesting above wouldn't need locking, it's just another structure on the stack line in this PR. Completely thread-local and call-frame-local.

Reply:
IMO the most common use case is to just call the get/set params function with a OSSL_PARAM. So iterating over the passed in params with a switch seems better for the most common use case, doesn't it?

Reply:
@levitte, for some reason I find the switch-if approach quite uncomfortable.  I also can't see the saved cost being significant -- it's generously a dozen array look-ups in total which is *far* better than a dozen strcmp().  My feeling for the performance limiting factor here will either be branch prediction misses or the `memset()` in `ossl_scan_params_for_idx()` -- it's 291 bytes in length currently.  What's a dozen memory reads (from cache) compared to that monster?

@hlandau, I think I understand what you are getting at.  That would address the large `memset()` for the addition of other complexity.  We'd essentially be using and switching on `params[indices[i]]` rather than `params[i]`.  This actually addresses a problem with the current code: duplicate params aren't handled properly & we do use duplicate params in places.

Reply:
Another option I am going to consider is building a separate lookup for each getter/setter function using only the specified params -- I doubt a TRIE is an ideal approach for this, rather a perfect hash would be the way (because it has less branch mispredictions).

I've also noticed that there are long strings of common letters in the middle of parameter names that could be squashed into a `strcasecmp()` call.  I'm not going to think about this now.

Reply:
> but doesn't that strike you as a lot of extra churn?

It's actually less churn the way I've done it, it should be an automatic process (or nearly) to convert existing code.  Changing to a loop-if is way more churn.

Reply:
> > but doesn't that strike you as a lot of extra churn?
> 
> It's actually less churn the way I've done it, it should be an automatic process (or nearly) to convert existing code. Changing to a loop-if is way more churn.

This, I just cannot see.

This PR has a solution where:

- `ossl_scan_params_for_idx()` *loops through the passed `OSSL_PARAM` array*
- calls `ossl_param_find_idx()` on each key and *stores away the result*
- the get/set function then has to *lookup the stored away result* (another, albeit comparatively very quick loop)

Comparing with the for-switch loop in the get/set function:

- It *loops through the passed `OSSL_PARAM` array*
- calls `ossl_param_find_idx()` on each key, and *uses the result directly*

I saw what you said about caches, and assume that's what you're on about in terms of churn.  I can't see, though, how the lookup of the `ossl_param_find_idx()` result makes it better in those terms than using the result directly.

Also, I personally find the for-switch loop in the get/set functions much more readable.

I did notice this,though:

> @levitte, for some reason I find the loop-switch approach quite uncomfortable. It precludes us processing parameters in the order we want to and I'm sure we've dependencies in this respect.

As far as I recall, it's when passing an algorithm name and a property query string as two separate parameters; we do some extra checks on those.  With a for-switch loop, those checks would have to be refactored, it's true.  The simplest would probably be to move such checks to happen after the for-switch loop, and doing so might involve a coupld of extra local variables, which I think would be quite ok, and possibly be more readable.  These combos are quite few, as I recall.

Reply:
I may have to pedal back re caching....  it didn't enter my mind that combining two separate switch statements in one loop might hit harder than I thought of.

Reply:
The cost for doing each lookup check is **trivial** once the array is populated.  It is literally:

```c
unsigned char idx[];

if (idx[3] != 0xff) { param found }
if (idx[6] != oxff) { param found }
...
```

I'm completely failing to see the _world shattering cost_ you seem to.  At most a dozen array index operations and integer compares.  This is nothing in the grand scheme.  That all memory accesses come from cache is a bonus.

It's a **far cry** from each being (param locate):
```
for (p=params; p->key!=NULL; p++)
   if (strcmp(p->key, key) == 0)
```
which was expensive.

Since I'm only seeing a few integer compares, could you please explain what I'm missing?


Note, I've also reworked this code to align with @hlandau's suggested for-switch loop.  I find this version less readable and maintainable personally.  However, stay tuned...there is an alternative path to explore still.

Reply:
Personally I like the for-switch arrangement. If we have an ordering issue with existing OSSL_PARAM usage I'd probably support reworking those specific usage sites to resolve it.

Reply:
I've been rethinking the cost for doing the original approach, setting the 291 bytes will likely cost less than a single branch miss prediction & quite possibly will occur during the misprediction stall time.  I.e. it's unlikely to make a meaningful difference & it avoids a fixed size on the passed param array.

Reply:
> ... & it avoids a fixed size on the passed param array.

Not sure I understand that bit...  when was that ever a case?

Reply:
@hlandau's approach requires an auxiliary array (of fixed size) and limit checks are required.

I've split this suggestion into a separate PR for comparison purposes.

Reply:
Aha ok, my misunderstanding then...  I read "the passed param array" as the `OSSL_PARAM` array the get/set functions receive.

Reply:
Poor choice of words on my part certainly helped :(